fem/intrules.cpp

```diff
@@ -1791,10 +1791,10 @@ IntegrationRule *IntegrationRules::TetrahedronIntegrationRule(int Order)
 IntegrationRule *IntegrationRules::PyramidIntegrationRule(int Order)
 {
    // This is a simple integration rule adapted from an integration
-   // rule for a cube which seems to be adequate for now. When we
-   // implement high order finite elements for pyramids we should
-   // revisit this and see if we can improve upon it.
-   const IntegrationRule &irc = Get(Geometry::CUBE, Order);
+   // rule for a cube which seems to be adequate for now. We should continue
+   // to search for a more appropriate integration rule designed specifically
+   // for pyramid elements.
+   const IntegrationRule &irc = Get(Geometry::CUBE, Order + 2);
```

Comment:
This increase in the order is actually not necessary, if the definitions of
* `IsoparametricTransformation::OrderW()`,
* `IsoparametricTransformation::OrderJ()`, and
* `IsoparametricTransformation::OrderGrad()`,

for `Uk` use the same values as `Qk`. One can treat integration on the reference pyramid with `Uk` spaces as integration on the reference cube with `Qk` spaces -- in this case the term `(1-z)^2` becomes part of the determinant of the Jacobian of the mapping from the reference cube (through the Duffy map to the reference pyramid) to the physical pyramid. The reason is that under the Duffy map `Uk` on the pyramid maps to a subset of `Qk` on the reference element.

Reply:
@v-dobrev, thanks again for your advice on these integration rules. I was able to use the lower integration order as you suggest here. I did need to bump up the order by 2 in both the CurlCurl and DivDiv integrators when these `Uk` function spaces are present. No other changes were necessary to get both the unit tests and the convergence tests to pass.